### PR TITLE
Add the metrics context to the metric creation API [PLEN-5]

### PR DESCRIPTION
**Pull request by nicu-da**
The labels provided during metric creation are merged with the labels provided during metric recording

### CODE REVIEW

File: ledger/metrics/src/main/scala/com/daml/metrics/api/dropwizard/DropwizardFactory.scala [hunk 1]
```diff
@@ -66,6 +75,9 @@ trait DropwizardFactory extends Factory {
 }
 
 trait FactoryWithDBMetrics extends DropwizardFactory {
+
+  def prefix: MetricName
```
Comment (nicu-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/dropwizard/DropwizardFactory.scala:79]: This was moved here as it was the only place where it was actually used

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 2]
```diff
@@ -141,39 +159,64 @@ case class OpentelemetryTimer(name: String, histogram: LongHistogram) extends Ti
...
+  def multiContextAsAttributes(context: MetricsContext*): Attributes = {
+    context
+      .foldLeft(Attributes.builder()) { case (builder, context) =>
+        context.labels.foreach { case (key, value) =>
+          builder.put(key, value)
+        }
+        builder
+      }
+      .build()
+  }
```
Comment (lucb-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:221]: It is a bit 'early optimisation', but I expect this function to be called multiple times for each commands ...

The first `case` is a partial function which is not needed, the function works just as well.
For the second `case`, `Map` has a function to avoid to do pattern matching.

```suggestion
  def multiContextAsAttributes(context: MetricsContext*): Attributes = {
    context
      .foldLeft(Attributes.builder()) { (builder, context) =>
        context.labels.foreachEntry { (key, value) =>
          builder.put(key, value)
        }
        builder
      }
      .build()
  }
```

Comment (lucb-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:221]: Personally, I might have gone for a `asAttributesWith(additional: MetricsContext) => Attributes` method on `MetricsContext`, as we don't really need to support a collection of `MetricsContext` for the moment.
But it is a matter of preferences.

**Review by lucb-da — COMMENTED** [review]
LGTM.
One comment on extra usage of `case`.

File: ledger/metrics/src/main/scala/com/daml/metrics/api/MetricHandle.scala [hunk 3]
```diff
@@ -19,24 +19,32 @@ object MetricHandle {
 
   trait Factory {
 
-    def prefix: MetricName
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/MetricHandle.scala:22]: Is there any reason behind removing `prefix` besides simplifying the `Factory` API?

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 4]
```diff
@@ -141,39 +159,64 @@ case class OpentelemetryTimer(name: String, histogram: LongHistogram) extends Ti
...
-  ): Unit = histogram.record(value.toLong, context.asAttributes)
+  ): Unit = histogram.record(
+    value.toLong,
+    AttributesHelper.multiContextAsAttributes(histogramContext, context),
+  )
+}
+
+private object AttributesHelper {
+
+  def multiContextAsAttributes(context: MetricsContext*): Attributes = {
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:212]: Can we elaborate the expected behaviour for attributes that are defined in multiple contexts, please?

**Review by miklos-da — COMMENTED** [review]
Can you add tests for merging metrics contexts, please?

File: ledger/metrics/src/main/scala/com/daml/metrics/api/Gauges.scala [hunk 5]
```diff
@@ -11,25 +11,18 @@ object Gauges {
...
-    ): Unit = ref.set(x -> context)
-    def updateValue(up: T => T)(implicit
-        context: MetricsContext
-    ): Unit = {
-      val _ = ref.updateAndGet { case (value, _) =>
-        up(value) -> context
+  case class VarGauge[T](initial: T) extends GaugeWithUpdate[T] {
+    private val ref = new AtomicReference[T](initial)
+    def updateValue(x: T): Unit = ref.set(x)
+    def updateValue(up: T => T): Unit = {
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/Gauges.scala:20]: Nit: `up` may be misleading here as it may imply 'upping'/increasing the value, hence, I'd suggest renaming it explicitly to `update` or simply `f` to improve readability.

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala [hunk 6]
```diff
@@ -57,62 +62,75 @@ trait OpenTelemetryFactory extends Factory {
...
+      context: MetricsContext = MetricsContext.Empty
+  ): MetricHandle.Histogram = OpentelemetryHistogram(
     name,
     otelMeter.histogramBuilder(name).ofLongs().build(),
+    context,
   )
 }
 
-case class OpentelemetryTimer(name: String, histogram: LongHistogram) extends Timer {
+case class OpentelemetryTimer(name: String, histogram: LongHistogram, timerContext: MetricsContext)
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:125]: Consistency: OpenTelemetry is spelled camel-cased, hence, I'd strongly suggest renaming all occurrences to follow that:
```suggestion
case class OpenTelemetryTimer(name: String, histogram: LongHistogram, timerContext: MetricsContext)
```
You may want to do this in a separate PR if it has a large impact radius.

File: ledger/metrics/src/test/suite/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactorySpec.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          )
+        ),
+      )
+      mergedAttributes.asMap().asScala should contain theSameElementsAs Map(
+        AttributeKey.stringKey("key1") -> "value",
+        AttributeKey.stringKey("key2") -> "value",
+      )
+    }
+
+    "merge contexts with duplicate label keys" in {
```
Comment (miklos-da) [inline ledger/metrics/src/test/suite/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactorySpec.scala:36]: :+1:

**Review by miklos-da — APPROVED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/api/Gauges.scala (hunk 5 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/api/Gauges.scala:20]: I would prefer that we do this in a separate PR as the API was actually defined in canton.

File: ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala (hunk 6 above)
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/api/opentelemetry/OpenTelemetryFactory.scala:125]: Please see already existing PR with this change: https://github.com/digital-asset/daml/pull/15357
